### PR TITLE
DPDK ring reader

### PR DESCRIPTION
**Pull request by jaroslavpesek**
Added a new input plugin, `dpdk-ring`, which enables ipfixprobe reading from shared memory as a consumer so that ipfixprobe can play the role of DPDK secondary process. See Examples in README.md for checking mandatory parameters.

### CODE REVIEW

**Comment by SiskaPavel** [comment]
Bude to chtit provest rebase

**Comment by jaroslavpesek** [comment]
rebasnuto

File: input/dpdk-ring.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+namespace ipxp {
+__attribute__((constructor)) static void register_this_plugin()
+{
+    static PluginRecord rec = PluginRecord("dpdk-ring", []() { return new DpdkRingReader(); });
+    register_plugin(&rec);
+}
+
+DpdkRingCore* DpdkRingCore::m_instance = nullptr;
+
+DpdkRingCore& DpdkRingCore::getInstance()
```
Comment (cejkato2) [inline input/dpdk-ring.cpp:63]:
```suggestion
DpdkRingCore *DpdkRingCore::m_instance = nullptr;

DpdkRingCore &DpdkRingCore::getInstance()
```

File: input/dpdk-ring.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    if (!m_ring) {
+        throw PluginError("Cannot find ring with name: " + parser.ring_name());
+    } else {
+        is_reader_ready = true;
+    }
+}
+
+struct timeval DpdkRingReader::getTimestamp(rte_mbuf* mbuf)
+{
+	struct timeval tv;
```
Comment (cejkato2) [inline input/dpdk-ring.cpp:165]:
```suggestion
    struct timeval tv;
```

File: input/dpdk-ring.h [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+     * @brief Configure DPDK secondary process.
+     * 
+     * @param eal_params DPDK EAL parameters.
+    */
+   void configure(const char* params);
+
+    /**
+     * @brief Get the singleton dpdk core instance
+     */
+    static DpdkRingCore& getInstance();
```
Comment (cejkato2) [inline input/dpdk-ring.h:112]:
```suggestion
    static DpdkRingCore &getInstance();
```

File: input/dpdk-ring.h [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+};
+
+class DpdkRingCore {
+public:
+    /**
+     * @brief Configure DPDK secondary process.
+     * 
+     * @param eal_params DPDK EAL parameters.
+    */
+   void configure(const char* params);
```
Comment (cejkato2) [inline input/dpdk-ring.h:107]:
```suggestion
   void configure(const char *params);
```

File: input/dpdk-ring.h [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    /**
+     * @brief Get the singleton dpdk core instance
+     */
+    static DpdkRingCore& getInstance();
+    void deinit();
+
+    DpdkRingOptParser parser;
+
+private:
+    std::vector<char *> convertStringToArgvFormat(const std::string& ealParams);
```
Comment (cejkato2) [inline input/dpdk-ring.h:118]:
```suggestion
    std::vector<char *> convertStringToArgvFormat(const std::string &ealParams);
```

File: input/dpdk-ring.h [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    void configureEal(const std::string& ealParams);
```
Comment (cejkato2) [inline input/dpdk-ring.h:119]:
```suggestion
    void configureEal(const std::string &ealParams);
```

File: input/dpdk-ring.h [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    ~DpdkRingReader();
+    DpdkRingReader();
+private:
+    std::vector<rte_mbuf*> mbufs_;
+    std::uint16_t pkts_read_;
+
+    void createRteMbufs(uint16_t mbufsSize);
+    struct timeval getTimestamp(rte_mbuf* mbuf);
```
Comment (cejkato2) [inline input/dpdk-ring.h:148]:
```suggestion
    struct timeval getTimestamp(rte_mbuf *mbuf);
```

File: input/dpdk-ring.h [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    std::string get_name() const override
+    {
+        return "dpdk-ring";
+    }
+
+    ~DpdkRingReader();
+    DpdkRingReader();
+private:
+    std::vector<rte_mbuf*> mbufs_;
```
Comment (cejkato2) [inline input/dpdk-ring.h:144]:
```suggestion
    std::vector<rte_mbuf *> mbufs_;
```

File: input/dpdk-ring.h [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    std::uint16_t pkts_read_;
+
+    void createRteMbufs(uint16_t mbufsSize);
+    struct timeval getTimestamp(rte_mbuf* mbuf);
+    DpdkRingCore& m_dpdkRingCore;
+    rte_ring* m_ring;
```
Comment (cejkato2) [inline input/dpdk-ring.h:150]:
```suggestion
    DpdkRingCore &m_dpdkRingCore;
    rte_ring *m_ring;
```

File: input/dpdk-ring.h [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    std::vector<char *> convertStringToArgvFormat(const std::string& ealParams);
+    void configureEal(const std::string& ealParams);
+    ~DpdkRingCore();
+    bool isConfigured = false;
+    static DpdkRingCore* m_instance;
+};
+
+class DpdkRingReader : public InputPlugin {
+public:
+    Result get(PacketBlock& packets) override;
```
Comment (cejkato2) [inline input/dpdk-ring.h:127]:
```suggestion
    Result get(PacketBlock &packets) override;
```

File: input/dpdk-ring.h [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    void deinit();
+
+    DpdkRingOptParser parser;
+
+private:
+    std::vector<char *> convertStringToArgvFormat(const std::string& ealParams);
+    void configureEal(const std::string& ealParams);
+    ~DpdkRingCore();
+    bool isConfigured = false;
+    static DpdkRingCore* m_instance;
```
Comment (cejkato2) [inline input/dpdk-ring.h:122]:
```suggestion
    static DpdkRingCore *m_instance;
```

**Review by cejkato2 — COMMENTED** [review]
There are some coding style issues.

License is not correct.

File: input/dpdk-ring.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+ * merchantability and fitness for a particular purpose are disclaimed.
+ * In no event shall the company or contributors be liable for any
+ * direct, indirect, incidental, special, exemplary, or consequential
+ * damages (including, but not limited to, procurement of substitute
+ * goods or services; loss of use, data, or profits; or business
+ * interruption) however caused and on any theory of liability, whether
+ * in contract, strict liability, or tort (including negligence or
+ * otherwise) arising in any way out of the use of this software, even
+ * if advised of the possibility of such damage.
+ *
```
Comment (cejkato2) [inline input/dpdk-ring.cpp:41]:
```suggestion
```

File: input/dpdk-ring.cpp [hunk 13]
```diff
@@ -0,0 +1,208 @@
+/**
+ * \file dpdk-ring.cpp
+ * \brief DPDK ring input interface for ipfixprobe (secondary DPDK app).
+ * \author Jaroslav Pesek <pesek@cesnet.cz>
+ * \date 2023
+ */
+/*
+ * Copyright (C) 2021 CESNET
```
Comment (cejkato2) [inline input/dpdk-ring.cpp:8]:
```suggestion
 * Copyright (C) 2023 CESNET
```

File: input/dpdk-ring.h [hunk 14]
```diff
@@ -0,0 +1,158 @@
+/**
+ * \file dpdk-ring.h
+ * \brief DPDK ring input interface for ipfixprobe (secondary DPDK app).
+ * \author Jaroslav Pesek <pesek@cesnet.cz>
+ * \date 2023
+ */
+/*
+ * Copyright (C) 2021 CESNET
```
Comment (cejkato2) [inline input/dpdk-ring.h:8]:
```suggestion
 * Copyright (C) 2023 CESNET
```

**Review by cejkato2 — CHANGES_REQUESTED** [review]

**Review by cejkato2 — APPROVED** [review]

**Pull request merged**